pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -187,11 +187,7 @@ func (gc *GarbageCollector) Sync(ctx context.Context, discoveryClient discovery.
 		}
 		if groupLookupFailures, isLookupFailure := discovery.GroupDiscoveryFailedErrorGroups(err); isLookupFailure {
 			// In partial discovery cases, preserve existing synced informers for resources in the failed groups, so resyncMonitors will only add informers for newly seen resources
-			for k, v := range oldResources {
-				if _, failed := groupLookupFailures[k.GroupVersion()]; failed && gc.dependencyGraphBuilder.IsResourceSynced(k) {
-					newResources[k] = v
-				}
-			}
+			newResources = gc.preserveFailedGroupResources(groupLookupFailures, oldResources, newResources)
```

Comment:
I actually find the encapsulation harder to reason about, and it being a method on gc makes me wonder if it is stateful